### ISSUE TITLE
Question: shared types between packages

### ISSUE THREAD

**Issue by Ashoat** [flow-typed/flow-typed#2842]
I help maintain the Flow libdefs for `react-navigation`. For our upcoming 3.0 release, we're planning on splitting the project into multiple npm packages. We'd ideally be able to define types that are shared across packages only once, but my understanding is that it's not yet possible to import types between libdefs.

My question is: what is the best practice for this today? Is copy-pasting types between libdefs the only way to do it? If so, does anybody know of a build system that automates/simplifies this process?

**Comment by Ashoat** [comment]
Prior discussions:

- https://github.com/flow-typed/flow-typed/issues/16
- https://github.com/flow-typed/flow-typed/issues/1795
- https://github.com/flow-typed/flow-typed/issues/1857

Flow issue:

- https://github.com/facebook/flow/issues/3883

**Comment by villesau** [comment]
Hi @Ashoat, thanks for asking!

TL;DR: Currently copy-pasting is indeed the best practice.

We have experimental branch that would support lerna monorepo structure for libdefs. The benefit is that it would allow to push all the libdefs to npm, and define cross dependecies between libdefs: https://github.com/flow-typed/flow-typed/tree/v3-experimental In addition to that CLI will need to be changed to support resolving libdef dependencies. cc @adityavohra7

**Comment by Ashoat** [comment]
So let's say we have libdefs for `react-navigation`, `react-navigation-stack`, and `react-navigation-tabs`. The idea is that we (React Navigation team) would create a lerna monorepo holding all three libdefs, and they would be able to cross-reference each other? And libdefs will no longer be hosted on this `flow-typed` repo, but instead published to npm?

**Comment by LoganBarnett** [comment]
@Ashoat are these projects for `react-navigation` going to use Flow directly? If so, we have some mechanisms to import the types defined in the source directly. It gets around the libdef problem. Generally we want to use  libdefs for libraries that don't have their own types defined.

**Comment by Ashoat** [comment]
No, they don’t use Flow directly.

**Comment by AndrewSouthpaw** [comment]
Also worth noting the experimental branch is far from finished, so I wouldn't bank on it any time soon. :-/

**Comment by Ashoat** [comment]
Wondering if there has been any update on this work. It's by far my #1 complaint about working with Flow...

**Comment by AndrewSouthpaw** [comment]
No progress has been made AFAIK, @Ashoat. 🙁

**Comment by Brianzchen** [comment]
Initial RFC [here](https://github.com/flow-typed/flow-typed/discussions/4242)
Implementation we plan to ship [here](https://github.com/flow-typed/flow-typed/pull/4331)

We're just working out the release plan atm as to not break users on current version

**Comment by Ashoat** [comment]
HUGE news, very exciting!! Thanks for the update and for making this happen, @Brianzchen :)

**Comment by Brianzchen** [comment]
@Ashoat I've just merged the v4 CLI which is to be published as an alpha initially but the test framework is fully functional and available on the `main` branch. Feel free to start raising PRs to that branch if you want to try out the new feature.

https://github.com/flow-typed/flow-typed/blob/main/CONTRIBUTING.md#importing-types-from-other-libdefs

**Comment by Ashoat** [comment]
Took a while to follow-up here, but I'm now working on updating the ReactNav libdefs for ReactNav6 and am excited to give this a whirl!! Will let you know if I have any trouble
